### PR TITLE
feat(examples): add Nature-DQN Pong training binary + runbook

### LINKED ISSUE #329: [Epic #306] Phase 4: reference-game training run to published-comparable score (operator-gated)
**Epic**: #306 — Atari (ALE) environment + CNN policy
**Phase**: 4 of 4
**Phase Goal**: Prove the stack end-to-end: train on the reference game to a published-comparable score.

## Description

Train DQN and/or PPO on the reference game chosen by the spike (#324) using the AtariEnv (#325/#326) and CNN policy (#327):

- Target: published-comparable score (if Pong: a positive score is decisive).
- Document the compute budget honestly — this is a long GPU run on alc-2, **operator-gated like #134**: the run itself needs operator hardware access and should not be auto-claimed by a builder without an operator arranging the run. The reproducible command, config, and learning-curve report are the reviewable artifacts.
- Report protocol: same as #298 (documented, reproducible command + honest learning-curve report), satisfying epic acceptance criterion #2.

## Acceptance Criteria

- [ ] Documented, reproducible training command (config committed)
- [ ] Learning-curve report with the honest budget (wall-clock, hardware, steps)
- [ ] Published-comparable score reached on the reference game (or an honest analysis of the gap)

## Dependencies

Part of Epic #306. Blocked by #325, #326, #327 (and benefits from #328's backend verdict for choosing the training backend).

---
*Decomposed from Epic #306*

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

Implements the **builder artifact** for Epic #306 Phase 4 (issue #329): a
committed Nature-DQN training binary for ALE Pong plus an operator runbook. The
long GPU run itself is operator-gated (alc-2) and executed post-merge — this PR
is the reviewable, CI-checkable artifact half of the artifact-vs-run split.

**No `src/` changes** — the full Atari/CNN/DQN stack is already on `main`. This
PR is one example binary, one `Cargo.toml` entry, and one runbook doc, matching
the curator's scope guard.

## Changes

- **`examples/games/atari/train_pong_dqn.rs`** (NEW): end-to-end Double-DQN
  trainer wiring `AtariPreprocess::new("pong", SEED)` (Machado defaults) →
  `NatureDqnQNetwork` → `DQNTrainerBurn`.
  - Compile-time backend selection via `#[cfg]` type alias: `cuda` →
    `Autodiff<Cuda<f32,i32>>`, `wgpu` → `Autodiff<Wgpu<f32,i32>>`, else
    `Autodiff<NdArray<f32>>` (CPU smoke path). Device built with
    `default_burn_device::<B>()`.
  - Forward closures reshape flat replay obs to `[B, 4, 84, 84]` for both the
    online and target passes; hard-copy target sync (Mnih-2015 spec, no soft tau).
  - Mnih-2015-adapted hyperparams (lr 2.5e-4 Adam, batch 32, buffer 100k,
    min-buffer 10k, target-sync 10k, ε 1.0→0.1 over 1M, γ 0.99, grad-clip 10).
  - Honest f32 replay-buffer memory math in the module doc (~21 GiB host RAM at
    100k; 210 GiB at the Mnih 1M — does not fit; 50k fallback).
  - Env knobs: `TOTAL_TIMESTEPS` (default 5M), `BUFFER_CAPACITY`,
    `MIN_BUFFER_SIZE`, `LOG_INTERVAL`, `CURVE_CSV`, `CHECKPOINT_INTERVAL`,
    `CHECKPOINT_DIR`, `ATARI_PYTHON`. Checkpoints via Burn's `BinFileRecorder`.
- **`Cargo.toml`**: one `[[example]]` entry, `required-features = ["training", "env-atari"]`.
- **`docs/PONG_DQN_RUNBOOK.md`** (NEW): exact alc-2 command sequence (pre-flight,
  tmux launch with CUDA, results locations), memory math, throughput
  expectation, and the local smoke-test recipe.

## Acceptance Criteria Verification (artifact criteria)

| Criterion | Status | Verification |
|-----------|--------|--------------|
| Example committed and compiles (`cargo check --features "training,env-atari"`) | ✅ | `cargo check --features "training,env-atari" --examples` passes |
| `[[example]]` entry with `required-features ["training","env-atari"]` | ✅ | Added after the pong examples in `Cargo.toml` |
| Smoke run passes CPU-side (worker connect, buffer fill, DQN updates, CSV, checkpoint, clean exit) | ✅ | Ran with a local ale-py 0.12 venv — see Test Plan |
| Runbook committed with alc-2 commands + memory math | ✅ | `docs/PONG_DQN_RUNBOOK.md` |
| `CURVE_CSV` + `CHECKPOINT_INTERVAL` documented in module doc | ✅ | Env-var knobs section of the module doc |

## Test Plan

Smoke run on this host (local `ale-py` 0.12 venv, `ATARI_PYTHON` pointed at it,
CPU NdArray backend, tiny budget):

```
TOTAL_TIMESTEPS=400 MIN_BUFFER_SIZE=100 BUFFER_CAPACITY=400 LOG_INTERVAL=50 \
CHECKPOINT_INTERVAL=200 CURVE_CSV=... ATARI_PYTHON=<venv>/bin/python3 \
./target/release/examples/train_pong_dqn
```

Confirmed: spawned the ale-py worker (Stella + Pong ROM loaded), filled the
buffer (50→400), ran 301 Double-DQN gradient updates after warmup **with no
shape panics**, wrote the CSV (header + 8 rows), saved `pong_dqn_200.bin` /
`pong_dqn_400.bin` (~6.7 MB each), and exited cleanly.

**Throughput signal for the alc-2 estimate**: CPU throughput is ~1 env-step/sec
after warmup (a full 1.6M-param CNN forward+backward per step on NdArray) — this
confirms the CPU path is smoke-only and the 5M-frame run needs the GPU, as
designed. On the RTX 4090 expect ~300–600 frame/s → ~2–5 h for 5M frames.

Verification:
- `cargo check --features "training,env-atari" --examples` ✅
- `cargo clippy --features "training,env-atari" --example train_pong_dqn -- -D warnings` ✅
- `cargo clippy --features "training,env-atari,cuda" --example train_pong_dqn -- -D warnings` ✅ (compile-only; no CUDA hardware here — the cuda **and** wgpu sets both `cargo check` cleanly on macOS)
- `cargo check --features "training,env-atari,wgpu" --example train_pong_dqn` ✅
- `cargo fmt --check` ✅
- `cargo doc --no-deps --features "training,env-atari"` (RUSTDOCFLAGS="-D warnings") ✅
- `cargo build` (default features) ✅ untouched

## Artifact-vs-run split

Per the #298 / epic #306 report protocol, the **run criteria** (positive Pong
score or honest gap analysis, learning-curve CSV, learning-curve report) are
fulfilled by the operator's alc-2 run and committed as a **follow-up** to
`docs/research/2026-07-pong-dqn-learning-curve.md` +
`docs/research/data/2026-07-pong_dqn_curve.csv`. This PR delivers the artifact
half only.

Closes #329

### CODE REVIEW

**Comment by rjwalters** [comment]
❌ **Changes Requested** — documentation-only frame-accounting fixes (no code change, no re-smoke needed)

The code is correct and well-built: Double-DQN is unconditional via `compute_td_target_double` in `src/train/dqn/trainer.rs`, the reshape to `[B,4,84,84]` is present in **both** the online/target forward closure (`forward_fn`) and the greedy action closure, hard target sync fires at `target_update_interval=10_000` via `self.target = self.online().clone()`, ε anneals 1.0→0.1 linearly over 1M steps, and all 10 CI checks pass (including the real-ale-py Atari integration job; both Burn NdArray test jobs green). Scope is clean: no `src/` changes, one minimal `[[example]]` entry, default builds untouched. `cargo check --features "training,env-atari" --example train_pong_dqn` passes locally.

**However, the frame-vs-step accounting in the docs is wrong in a way that misleads the operator, and the runbook's timing estimate is internally contradictory. These are the blocking items — all documentation-only.**

### 1. Wrapper steps ≠ frames (frame-skip 4)

`AtariPreprocess::step()` applies **frame-skip 4** (`src/env/games/atari/preprocess.rs`: the frame-skip loop submits the action to the inner ALE 4× per wrapper step). The training loop counts **wrapper steps** (`trainer.total_env_steps()`), and `TOTAL_TIMESTEPS=5_000_000` is a **wrapper-step** budget. So:

- **5M wrapper steps = 20M raw ALE frames**, not "5M frames".

The module doc (`examples/games/atari/train_pong_dqn.rs`) and runbook conflate these:
- Example line 24 "5M-frame run", line 56 "≤10M-frame budget", line 58 table header "This run (5–10M)", lines 71–74 "targets ≤10M frames".
- Runbook lines 16–18, 37, 39, 52, 58.

The Mnih literature convention you're comparing against ("50M frames", "Pong crosses zero well within 10M frames") counts **raw frames**. Mixing the two conventions makes the budget claim ambiguous. Please pick one convention and label it explicitly everywhere. Concretely, the honest framing is: *5M wrapper steps = 20M raw frames, which is ~2× the "≤10M raw frames" the spike analysis says Pong needs to cross zero* — so the Pong-positive claim is actually **safe/conservative**, but only once the frames-vs-steps units are stated. Right now a reader can't tell whether "5M" is under or over the "10M frames" bar.

### 2. Runbook throughput estimate is internally contradictory and ~4× optimistic (`docs/PONG_DQN_RUNBOOK.md` lines 113–122)

- Line 117–118: "~300–600 frame/s (i.e. ~75–150 ALE steps/s given frame-skip 4)". If 300–600 is *raw frames/s*, then wrapper-steps/s = 75–150 — and the `fps=` field the code logs is **wrapper steps/sec** (`total_env_steps / elapsed`, line 304 of the example). So the operator will see `fps=75..150`, not 300–600.
- Line 121: "a 5M-frame run is roughly **2–5 hours** at that rate." But the loop runs to **5M wrapper steps**. At 75–150 wrapper steps/s that is ~9–18 h, not 2–5 h. The 2–5 h figure only holds if the loop did 300–600 *wrapper* steps/s, which contradicts the "frame-skip 4" note two lines up.

Please make this self-consistent: state the expected `fps=` value in the **same unit the code prints (wrapper steps/s)**, then derive raw frame/s = 4× that, then derive wall-clock from `5M wrapper_steps / (wrapper_steps_per_sec)`. Whatever the true 4090 rate is, the three numbers (`fps=`, raw frame/s, hours) must agree.

### Non-blocking notes (fix if convenient; not required for approval)

- **Reward clipping**: the Mnih-2015 hyperparameter table implies fidelity but omits the reward-clipping (±1) row. The trainer does **not** clip (`src/train/dqn/loss.rs` passes rewards through raw). For Pong this is a behavioral no-op (per-point reward is already ±1), so it doesn't affect the run — but a one-line note in the table ("reward clipping: n/a for Pong, rewards already ±1") would make the honesty complete.
- **Env-override robustness**: `env_usize()` (example line 377) silently falls back to the default on unparseable input (`.parse().ok()...unwrap_or(default)`). A typo like `TOTAL_TIMESTEPS=5000000x` would silently run the 5M default rather than erroring — surprising for an operator-launched multi-hour run. Consider `expect`/`bail!` on a present-but-unparseable value. `CHECKPOINT_INTERVAL` has the same silent-fallback shape.

Once the frame-vs-step units are consistent across the module doc + runbook and the throughput/timing figures reconcile, I'll approve — this is purely a docs pass, so no code re-test is needed.

**Comment by rjwalters** [comment]
✅ Documentation-only frame-accounting fixes addressed (commit 83c30c9). No behavioral change to the training loop.

**1. Frame-vs-wrapper-step units (blocking) — fixed.** Stated the convention explicitly everywhere in both `examples/games/atari/train_pong_dqn.rs` (new "Frames vs. wrapper steps" section + updated backend note, hyperparameter table header/column, Mnih prose, and the `TOTAL_TIMESTEPS` knob doc) and `docs/PONG_DQN_RUNBOOK.md` (success criterion, hyperparameter table, env-var table). Consistent framing throughout: **5M wrapper steps = 20M raw ALE frames (frame-skip 4)**, which is ~2× the ~10M raw frames the spike analysis says Pong needs to cross zero — so the Pong-positive claim is conservatively over-budgeted, and now says so.

**2. Throughput block self-contradiction (blocking) — fixed.** Rewrote the runbook throughput section around a single convention: the code's `fps=` field is **wrapper steps/s**. Added a table giving, per case, `fps=` (wrapper steps/s) → raw frames/s (×4) → wall-clock for 5M wrapper steps:
- Optimistic (GPU-bound): `fps=`300–600 → 1200–2400 raw f/s → ~2.5–5 h
- Conservative (IPC-bound): `fps=`75–150 → 300–600 raw f/s → ~9–18 h

and noted that the CPU smoke run can't predict which case the run lands in; operator derives wall-clock from the observed `fps=`.

**3. Non-blocking niceties — both done.**
- Added the reward-clipping row (`±1` Mnih / `n/a for Pong (rewards already ±1)`) to both Mnih-fidelity tables.
- `env_usize()` and `CHECKPOINT_INTERVAL` now **fail loudly** on a present-but-unparseable value (`anyhow` error) instead of silently reverting to the default. Unset still uses the default; `0` still disables checkpointing.

**Verification:**
- `cargo check --features "training,env-atari" --example train_pong_dqn` — clean.
- `cargo clippy --features "training,env-atari" --example train_pong_dqn -- -D warnings` — clean.
- `cargo fmt` — applied, clean.
- Standard `cargo doc` build — clean (the 4 `-D warnings` redundant-intra-doc-link errors are **pre-existing on the base branch**, untouched module-doc lines 5/6/8; confirmed by stashing my changes).
- Env-parsing change re-verified against the binary: `TOTAL_TIMESTEPS=abc` and `TOTAL_TIMESTEPS=5000000x` and `CHECKPOINT_INTERVAL=xyz` now each error clearly and exit before any ALE setup; a valid `TOTAL_TIMESTEPS=200` parses through and only fails at the ALE worker launch — proving the happy path is intact. (No live `ale-py` interpreter is present on this fixer host, so the full worker→buffer→CSV→checkpoint smoke couldn't be executed here, but the change is confined to the top-of-`main` parse layer; the loop below is unmodified.)

Ready for re-review.

**Comment by rjwalters** [comment]
✅ **Approved!** (Re-review after Doctor pass, commit `83c30c9`)

The prior rejection was documentation-only. I verified every item in the post-doctor diff against the actual sources and the built binary. All three blocking items and both non-blocking items are resolved, and full CI is green.

## Units — now explicit and consistent everywhere
- The frame-vs-wrapper-step conflation is gone. Both the module doc (`# Frames vs. wrapper steps`) and the runbook (`Units — frames vs. wrapper steps`) state the invariant plainly: **frame-skip 4 → 5M wrapper steps = 20M raw frames**. `TOTAL_TIMESTEPS` is labeled a wrapper-step budget consistently in the success criterion, both hyperparam tables, and both env-var tables.
- Mnih comparison is honest: 20M raw frames stated as **≈ 2× the ~10M raw-frame "cross zero" bar** from the spike analysis — framed as conservatively over-budgeted for a *positive* score, not a ceiling score. Baselines (DQN ≈ 18.9 @ 50M raw frames) are consistently in raw-frame units.

## Throughput block — one convention, no contradiction
- The runbook now fixes a single convention up front: `fps=` logs **wrapper steps/s** (`total_env_steps / elapsed`), with `raw frames/s = 4 × fps` and `wall-clock = 5M / fps`. The two-row table (optimistic GPU-bound ~2.5–5 h; conservative IPC-bound ~9–18 h) is internally consistent and matches the worked examples (`fps=100 ⇒ ~14 h`, `fps=400 ⇒ ~3.5 h`). The self-contradiction is gone.

## Reward-clipping row + fail-loud env parsing (the non-blocking items)
- Reward-clipping row present in both hyperparam tables: `reward clipping | ±1 | n/a for Pong (rewards already ±1)`.
- Env parsing now fails loudly. Verified against the built binary (release, `training,env-atari`):
  - `TOTAL_TIMESTEPS=abc` → `Error: TOTAL_TIMESTEPS="abc" is not a valid usize: invalid digit found in string`, exits before spawning the worker.
  - `CHECKPOINT_INTERVAL=xyz` → analogous clear error.
  - `env_usize` falls back to the default **only when the var is unset** (`Err(_) => Ok(default)`); present-but-unparseable is a hard error. The training loop below the parse layer is unchanged from the first-pass-verified code (Double-DQN, NCHW reshapes, hard target sync, ε schedule, hyperparams).

## The 4 intra-doc-link references the Doctor flagged
- These are the `[\`NatureDqnQNetwork\`](thrust_rl::...)`-style links in the example's module doc. I confirmed they are **CI-invisible and non-blocking**:
  - The Documentation job runs `cargo doc --no-deps --all-features` (RUSTDOCFLAGS=`-D warnings`), which does **not** document files under `examples/`. Reproduced locally: exit 0, no warnings — and CI's Documentation job is green on `83c30c9`.
  - They live in an unpublished example (`examples/` is excluded from the crate tarball; the publish-audit job is green), don't affect `cargo check`/`clippy`/build, and would only surface under a stricter `cargo doc --examples`-style build this repo doesn't run.
  - They are cosmetic and do not meet the bar for a blocking defect on an exhausted-Doctor-cycle PR.

## CI — fully green on 83c30c9
All 10 checks pass, including the authoritative jobs:
- Atari integration (real ale-py) ✅, Documentation ✅, Security Audit ✅
- Tests (Linux + Burn NdArray) ✅, Tests (macOS + Burn NdArray) ✅
- Clippy ✅, Rustfmt ✅, Check (no training) ×2 ✅, Publish dry-run + tarball audit ✅
- `mergeStateStatus`: CLEAN

## Local verification performed
- `cargo doc --no-deps --all-features` (RUSTDOCFLAGS=`-D warnings`) → clean (reproduces CI docs job; confirms example doc not covered).
- Built the example (`--release --features "training,env-atari"`) → OK; exercised the fail-loud parse layer as above.

PR description uses `Closes #329`. Great cleanup by the Doctor — the docs are now unit-honest end to end. Shipping it.
